internal/pkg/infrastructure/postgres/readingagg.go

```diff
@@ -100,14 +151,98 @@ func (c *Client) ReadingsAggregationByDeviceNameAndResourceName(deviceName strin
 }
 
 // ReadingsAggregationByDeviceNameAndResourceNameAndTimeRange queries aggregated reading values by device name & resource name within the given time range using the specified SQL aggregation function.
-func (c *Client) ReadingsAggregationByDeviceNameAndResourceNameAndTimeRange(deviceName string, resourceName string, aggregateFun string, start int64, end int64) ([]model.Reading, errors.EdgeX) {
-	sqlStatement := sqlQueryAggregateReadingWithConds(aggregateFun, true, deviceNameCol, resourceNameCol)
-
-	readings, err := queryReadings(context.Background(), c.ConnPool, sqlStatement,
-		pgx.NamedArgs{deviceNameCol: deviceName, resourceNameCol: resourceName, startTimeCondition: start, endTimeCondition: end})
+func (c *Client) ReadingsAggregationByDeviceNameAndResourceNameAndTimeRange(deviceName string, resourceName string, aggregateFunc string, start int64, end int64, offset int, limit int) ([]model.Reading, errors.EdgeX) {
+	sqlStatement := sqlQueryAggregateReadingWithCondsAndPag(aggregateFunc, true, deviceNameCol, resourceNameCol)
+	offset, validLimit := getValidOffsetAndLimit(offset, limit)
+
+	readings, err := queryAggReadings(
+		context.Background(),
+		c.ConnPool,
+		sqlStatement,
+		pgx.NamedArgs{deviceNameCol: deviceName, resourceNameCol: resourceName, startTimeCondition: start, endTimeCondition: end, offsetCondition: offset, limitCondition: validLimit},
+		aggregateFunc,
+	)
 	if err != nil {
 		return nil, errors.NewCommonEdgeX(errors.KindDatabaseError,
 			fmt.Sprintf("failed to query readings by device '%s' and resource '%s'", deviceName, resourceName), err)
 	}
 	return readings, nil
 }
+
+// queryAggReadings queries the aggregate readings data rows with given sql statement and passed args,
+// converts the rows to map and unmarshal the data rows to the Reading model slice
+func queryAggReadings(ctx context.Context, connPool *pgxpool.Pool, sql string, args pgx.NamedArgs, aggregateFunc string) ([]model.Reading, errors.EdgeX) {
+	rows, err := connPool.Query(ctx, sql, args)
+	if err != nil {
+		return nil, pgClient.WrapDBError("query failed", err)
+	}
+
+	readings, err := pgx.CollectRows(rows, func(row pgx.CollectableRow) (model.Reading, error) {
+		readingDBModel, err := pgx.RowToStructByNameLax[dbModels.Reading](row)
+		if err != nil {
+			return nil, pgClient.WrapDBError("failed to convert row to map", err)
+		}
+
+		// convert the BaseReading fields to BaseReading struct defined in contract
+		baseReading := readingDBModel.GetBaseReading()
+
+		reading := model.NumericReading{
+			BaseReading: baseReading,
+		}
+		if readingDBModel.NumericValue != nil {
+			// reading type is NumericReading
+			val, err := parseAggNumericReading(readingDBModel.ValueType, aggregateFunc, readingDBModel.NumericValue)
```

Comment:
modify the valueType from the reading model as well

Reply:
Done